### PR TITLE
Create pid-file atomically

### PR DESCRIPTION
**Pull request by crosbymichael**
For things that depend or watch for this pid file to know when the
container is started we need to create this file atomically.

Signed-off-by: Michael Crosby <crosbymichael@gmail.com>

### CODE REVIEW

**Comment by LK4D4** [comment]
LGTM

File: utils.go [hunk 1]
```diff
@@ -269,18 +269,27 @@ func setupIO(process *libcontainer.Process, rootuid int, console string, createT
 	return createStdioPipes(process, rootuid)
 }
 
+// createPidFile creates a file with the processes pid inside it atomicly
```
Comment (mrunalp) [inline utils.go]: typo: atomicly

File: utils.go [hunk 2]
```diff
@@ -269,18 +269,27 @@ func setupIO(process *libcontainer.Process, rootuid int, console string, createT
...
 	if err != nil {
 		return err
 	}
 	defer f.Close()
-	_, err = fmt.Fprintf(f, "%d", pid)
-	return err
+	if _, err := fmt.Fprintf(f, "%d", pid); err != nil {
+		return err
+	}
+	return os.Rename(tmpName, path)
```
Comment (mrunalp) [inline utils.go]: should we close the file before the rename?

Comment (crosbymichael) [inline utils.go]: not sure but would probably be safer

Comment (mrunalp) [inline utils.go]: yeah, agree

**Comment by mrunalp** [comment]
LGTM

**Comment by hqhq** [comment]
@crosbymichael Can you explain what's the problem if we create the pid file directly?

File: utils.go [hunk 3]
```diff
@@ -269,18 +269,28 @@ func setupIO(process *libcontainer.Process, rootuid int, console string, createT
...
+// then renames the file
 func createPidFile(path string, process *libcontainer.Process) error {
 	pid, err := process.Pid()
 	if err != nil {
 		return err
 	}
-	f, err := os.OpenFile(path, os.O_RDWR|os.O_CREATE|os.O_EXCL, 0666)
+	var (
+		tmpDir  = filepath.Dir(path)
+		tmpName = filepath.Join(tmpDir, fmt.Sprintf(".%s", filepath.Base(path)))
```
Comment (marcosnils) [inline utils.go]: I'm a bit late for this, I'm just wondering if there's a reason why TemFile(https://golang.org/pkg/io/ioutil/#TempFile) is not being used instead.

**Pull request merged**